torch/_dynamo/symbolic_convert.py

```diff
@@ -949,6 +949,61 @@ def COMPARE_OP(self, inst):
                     **options,
                 )
             )
+        elif (
+            isinstance(left, BaseListVariable)
+            and all(
+                isinstance(l, (ConstantVariable, DynamicShapeVariable))
+                for l in left.items
+            )
+            and isinstance(right, BaseListVariable)
+            and all(
+                isinstance(r, (ConstantVariable, DynamicShapeVariable))
+                for r in right.items
+            )
+            and len(left.items) == len(right.items)
+            and op in supported_tensors
```

Comment:
I think we want to do this recursively. If we have two lists, AND we've specialized on their sizes (I think this is always true for ListVariable?) we want to recursively run comparison on each of their elements. What we essentially have done is reimplemented the `__eq__` implementation of the list itself, because that's ostensibly in CPython C itself (specifically, `PyObject_RichCompare`), however, the *inside* should be delegated to whatever other knowledge we have.

Reply:
This does kind of suggest we need some sort of "tp_richcompare" concept on VariableTracker, maybe, to keep track of this sort of thing. cc if @jansel has any thoughts

Reply:
Doing it recursively sounds good to me. Although speaking of this comment:

> This is wrong, you'll report true if you compare a list and a tuple but these are never equal

Obviously I could include a check that they're the same type here, but I agree there are an annoying amount of checks here. Is there a way that you can think of to avoid having dynamo add a manual check like that?

Reply:
The "proper" way to do this is to replicate the logic that happens in CPython (which I am not entirely sure how it works). Something quick like only handling if the type() is the same, that seems fine (because it's always sound to unsupported otherwise)

Reply:
I think we could just add a `compare()` method to `VariableTracker` that raises an exception for unhandled.  I think a lot of the complexity would be reduced if we used method dispatch instead of `isinstance`.  It would also make the recursive check way easier.  For example, BaseListVariable could just check type and length then recursively call compare on all the items.

Reply:
+1 this should be on VariableTracker